### ISSUE TITLE
Memory Limit when calling "Declare_on_GPU"

### ISSUE THREAD

**Issue by nilsmelchert** [Jerome-LMA/oscar#11]
Hi,

I experienced Memory limitations of the GPU, when using the method `Declare_on_GPU` in the branch `next_release`.

As pointed out in my GPU-Issue #1, I was able to calculate the FSR-Scan for up to 512x512 EFields. However, when preallocating GPU-Memory with `Declare_on_GPU` inside of the Cavity1 class, there is not sufficient memory left for high resolution scans.

I think, it would make sense to work with flags, as you already declared in the Cavity1 class:
https://github.com/Jerome-LMA/oscar/blob/79e8ad4b6ea45322d810054ba75c5f18af348450/Classes/%40Cavity1/Cavity1.m#L16

And then simply convert the Arrays to GPU-Arrays during runtime on demand. I am pretty sure, that we will not lose that much performance with that and can also calculate FSR-Scans with higher resolutions.

**Comment by nilsmelchert** [comment]
I am currently still investigating this with respect to performance and keep this issue up to date.

**Comment by nilsmelchert** [comment]
I did some tests of FSR-Scan calculation initializing the `Cavity1` class with and without the new `Declare_on_GPU` method.

With calling `Declare_on_GPU`:
![image](https://user-images.githubusercontent.com/36985893/154150698-798ed242-ce21-4b01-ba4b-e24507e0f4fc.png)

Without calling `Declare_on_GPU`:
![image](https://user-images.githubusercontent.com/36985893/154150772-e7315a1a-65af-4950-8c32-459c27397e54.png)

Since the FFT of the EFields is a relatively small amount of data, it is faster to calculate this on the CPU than on the GPU. The time it takes to send the data to the GPU memory takes a relatively long time.

I would propose, that we use GPU-Arrays on demand and not initialize them everywhere we can, because we might lose performance and memory that way.


What do you think @Jerome-LMA ?

**Comment by Jerome-LMA** [comment]
About the memory issue, I usually would not recommend to do FSR scan with a grid size about 256 x 256, it becomes very slow and such resolution is seldom needed. So I would keep there a warning. A complex matrix of 256x256x200, yes could be huge in size.

I will try to reproduce your results to understand better the right strategy, I was expecting a gain to run on the GPU.

**Comment by Jerome-LMA** [comment]
@nilsmelchert could you try again with the updated version of the branch 'Next Release', I fixed an issue. Thanks!

**Comment by nilsmelchert** [comment]
@Jerome-LMA Yes, I will do the performance evaluation today and post the results here afterwards.
